Add tensor_mechanics version of XFEMMaterialTensorMarkerUserObject

bwspenc opened the pull request:
ref #9604

bwspenc committed:
Add tensor_mechanics version of XFEMMaterialTensorMarkerUserObject ref #9604

It's called XFEMRankTwoTensorMarkerUserObject. Also add multiple tests for this
capability.

To support this, enable calculation of directions for principal stresses in
RankTwoScalarTools.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/105691/) on b70c129 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format d6812abd5afd5772250c15f60a868d286aadb0b3`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9616/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9616/style.patch | git apply -v`

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/105756/) on 93258b3 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9616/site)

jiangwen84 commented:
It looks good 👍

jiangwen84 reviewed (commented):
It looks good 👍

permcody commented on modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C:
```diff
@@ -0,0 +1,110 @@
+          _tensor[qp], _scalar_type, _point1, _point2, _q_point[qp], directions[qp]);
+      if (directions[qp](0) == 0 && directions[qp](1) == 0 && directions[qp](2) == 0)
+        mooseError("Direction has zero length in XFEMRankTwoTensorMarkerUserObject");
+      if (tensor_quantities[qp] > max_quantity)
```
Minor: Is it possible for _any_ user selected tensor_quantities value to be negative? My guess is no, but if it ever is, this branch won't be entered, which under a worst case scenario could cause you to access directions[numeric_limists<unsigned int>::max()] in the next branch below (if _threshold was also poorly chosen). You are clearly making at least one range assumption here, it probably should be asserted. Also, you might consider a range check on _threshold as well.

bwspenc replied:
@permcody Good point.  I can think of cases where you might want the threshold to be negative, so I revised the code accordingly.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/106350/) on 9ac846a wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format d6812abd5afd5772250c15f60a868d286aadb0b3`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9616/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9616/style.patch | git apply -v`

bwspenc commented:
Oh yeah, documentation... I'll work on that!

bwspenc commented:
@jasondhales I think this is ready to go (assuming the tests pass).  I addressed the issue pointed out by @permcody by enforcing that the threshold is always positive. The TM and SM versions of this test use a common gold file, and I added a few extra variants of the test for TM.

moosebuild commented:
Job [External App tests](https://www.moosebuild.org/job/108425/) on 819c461 : invalidated by @bwspenc

It looks like this successfully completed, but it&#39;s not showing up as being done

bwspenc commented:
Now that I've resolved the test issues, I think this is good to go.

moosebuild commented:
Job [Test Intel](https://www.moosebuild.org/job/108899/) on f6722ff : invalidated by @bwspenc

I&#39;m not sure why this was cancelled. Resubmitting.

bwspenc commented:
Does anyone know what's up with the 'Test Intel' tests being cancelled? @brianmoose @permcody? I tried invalidating it, but that didn't seem to do anything.

There was an issue with this commit earlier today -- a conflict because of the change to the test directories. That has been straightened out. Aside from the Intel tests not running, I think this is ready to go.

permcody commented:
We've been having several problems with the testing infrastructure today. Not only have we run out of disk space on some of the clients (a somewhat regular occurrence), but we've also had a few networking and load issues with the CIVET server. We are looking into the later now. No need to do anything, we'll sort it all out. The TigerTeam activity has pushed the systems to their limits several times over the past 24 hours.

permcody commented on modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C:
```diff
@@ -0,0 +1,114 @@
+
+  if (_average)
+  {
+    Real average_threshold;
```
You aren't initializing this Real! The first assignment is a `+=` which could give you garbage.

bwspenc replied:
Doh!  I'll fix that.

permcody commented on modules/xfem/src/userobjects/XFEMRankTwoTensorMarkerUserObject.C:
```diff
@@ -0,0 +1,114 @@
+        mooseError("Threshold must be strictly positive in XFEMRankTwoTensorMarkerUserObject");
+      const Real tensor_quantity = RankTwoScalarTools::getQuantity(
+          _tensor[qp], _scalar_type, _point1, _point2, _q_point[qp], directions[qp]);
+      if (directions[qp](0) == 0.0 && directions[qp](1) == 0.0 && directions[qp](2) == 0.0)
```
Maybe use `directions[qp].size() == 0.0` or a fuzzy comparison?

bwspenc replied:
Sure.  I'll do that.  These really will be set to zero in this case, but I'll make it fuzzy.

permcody reviewed (changes requested):
I spotted one potential bug. Can you take a look?

bwspenc commented:
@permcody I addressed your requested changes.
